cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+//
+// Copyright 2019 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/descriptor"
+	"github.com/golang/protobuf/proto"
+	filedescr "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"log"
+	"net"
+	"os"
+	"runtime"
+)
+
+const (
+	SupportLibraryVersion = "0.1.0"
+	SupportLibraryName    = "cloudstate-go-support"
+)
+
+// CloudState is an instance of a CloudState User Function
+type CloudState struct {
+	server                   *grpc.Server
+	entityDiscoveryResponder *EntityDiscoveryResponder
+	eventSourcedHandler      *EventSourcedHandler
+}
+
+// NewCloudState returns a new CloudState instance.
+func NewCloudState(options *Options) *CloudState {
+	cs := &CloudState{
+		server:                   grpc.NewServer(),
+		entityDiscoveryResponder: NewEntityDiscoveryResponder(options),
+		eventSourcedHandler:      NewEventSourcedHandler(),
+	}
+	protocol.RegisterEntityDiscoveryServer(cs.server, cs.entityDiscoveryResponder)
+	log.Println("RegisterEntityDiscoveryServer")
+	protocol.RegisterEventSourcedServer(cs.server, cs.eventSourcedHandler)
+	log.Println("RegisterEventSourcedServer")
+	return cs
+}
+
+// Options go get a CloudState instance configured.
+type Options struct {
+	ServiceName    string
+	ServiceVersion string
+}
+
+var NoOptions = Options{}
+
+// DescriptorConfig configures service and dependent descriptors.
+type DescriptorConfig struct {
+	Service        string
+	ServiceMsg     descriptor.Message
+	Domain         []string
+	DomainMessages []descriptor.Message
+}
```

Comment:
- [ ] comment fields in DescriptorConfig

Reply:
>Can we replace this struct and implementation with invoking Descriptor() directly on the Message to get the descriptors?

Good point and also one I iterated for some time. This type is used with `cloudState.Register` where a user registers an event sourced entity like so:

```
err := cloudState.Register(
		&cloudstate.EventSourcedEntity{
			Entity:      (*ShoppingCart)(nil),
			ServiceName: "com.example.shoppingcart.ShoppingCart",
		},
		cloudstate.DescriptorConfig{
			Service: "shoppingcart/shoppingcart.proto",
		}.AddDomainDescriptor("domain.proto"),
	)
```

Fo the Service, like in the Java Implementation, I would preferably use a type safe reference to a (Service)Descriptor. The Go gRPC implementation does not export them unfortunately. So to get the Service and any other dependent descriptors the user either provides their names or any of the messages used for the service or those dependent messages.

For the shopping cart, this would be one of the messages used like `shoppingcart.AddLineItem` or `shoppingcart.Cart` which I found irritating from a users perspective.

The `DescriptorConfig` in this terms lets the user choose how to register the service and dependent descriptors. I see that this is not optimal and gives too many degrees of freedom from the users perspective.

I repeat here how that would look like:

Cart
```
		cloudstate.DescriptorConfig{
			Service: shoppingcart.AddLineItem,
		}.AddDomainDescriptor(domain.ItemAdded),
```

Do you see any other ways to configure a gRPC Service and their dependent descriptors @jsravn ?

Reply:
I found https://github.com/jhump/protoreflect, which seems to support retrieving the descriptors from the linked in pb.go files. Although I haven't tested it myself. It also has logic for loading directly from proto files, which we could use instead of implementing it ourselves.

Reply:
https://godoc.org/github.com/jhump/protoreflect/desc#FileDescriptor.FindService for example.

Reply:
I have seen protoreflect and I had the impression I could not get the descriptor this way. It also seems quite a big dependency to depend on tho.